text/0232-maturity-of-otel.md

```diff
@@ -0,0 +1,151 @@
+
+#### Development
+
+Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production.
+
+#### Alpha
+
+This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The configuration options might often change without backward compatibility guarantees.
+
+#### Beta
+
+Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.
+
+#### Stable
+
+The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
+
+#### Deprecated
+
+Development of this component is halted, and no further support will be provided. No new versions are planned, and the component might be removed from its included distributions. Note that new issues will likely not be worked on except for critical security issues. Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed.
+
+#### Unmaintained
+
+A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.
+
+### Specification requirements
+
+The specification should declare all requirements for a specified component to be declared stable. When implementations are still missing features required by the specification, they can be at most Alpha status.
```

Comment:
I think this would put .NET in Alpha, as it does not implement the Context spec and also has some other incompatibilities, e.g. https://github.com/open-telemetry/opentelemetry-dotnet/pull/3841. As far as I understand, this is due to it not providing a full OTel implementation but cooperating with the .NET runtime for providing required features. Probably a special exception is needed here.

Reply:
cc @open-telemetry/dotnet-approvers

Reply:
Agreed, an alpha designation does not make sense for .NET, but we are a unique API/SDK in some ways.

I haven’t taken a pass over the spec compliance matrix in some time, and it's possible it may need some updates.

I take some time next week to take a look at the matrix for .NET, submit a PR with updates as necessary, and review it with an eye for what our suggestion should be regarding this OTEP.

My thought off hand is that perhaps we should suggest a way to accurately reflect in the matrix deviations w.r.t. spec compliance that we do not consider affecting the SDK's maturity.

Reply:
We could have a provision here for exceptions, but I'm curious as to the message we send when we state that "Context" is required and .NET doesn't implement it.

Reply:
I would for sure prefer it .NET implemented Context... But not sure if it is realistic. Or we could give .NET a N/A designation?

Reply:
Based on the Slack conversation, my understanding is that the parts of the spec that are not implemented by dotnet are optional only. No required aspects are left unimplemented.

Reply:
I've responded in the Slack thread. Summarizing here as well.

With regards to the Context spec, I think .NET's [`RuntimeContext`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/RuntimeContext.cs) covers some of the concerns. There are other parts of the spec related to context that .NET has not implemented, e.g., span processor OnStart does not have an explicit context parameter. Though, RuntimeContext may be sufficient for many use cases.

Reply:
I'm unresolving this because I don't think RuntimeContext covers any of the concerns as long as it's not used anywhere at all by the SDK and API, not even in the places it's required 😃

Reply:
> I'm just saying that it technically does not fit the current stability definition.

Agreed. You have noted a couple examples where .NET does not comply with the spec, but I think the general point you raise is valid:

1. How do we evaluate stability/maturity?
2. Given that .NET maintainers have declared the API/SDK stable despite some deviations from the spec, is the maintainers' judgement sufficient to declare stability?
3. If it Is it possible for an API/SDK to be declared stable despite deviations from the spec, should we have a means to denote/identify deviations in the spec compliance matrix? If so, it would seem that the `+/-` designations in the matrix is insufficient. `-` seems to suggest that something just hasn't been implemented yet. `+` should probably mean it is 100% compliant with the spec. For items not 100% compliance, maybe we need to denote it either `N/A` (not going to implement) or `+*` (implemented but deviates w/ info captured in a footnote).

Reply:
Thank you all for the clarifications. I'll split this discussion into its own issue, especially after today's decision to get this OTEP to focus on the maturity level framework, leaving the concept of tier-1 and the list of tier-1 SIGs out for now.

Reply:
I was confident I had, but I can't find it. I just opened one, to handle the generic case: https://github.com/open-telemetry/opentelemetry-specification/issues/3673